[Refactor, Fix]: Agent controller state/metrics management

malhotra5 opened the pull request:
- [ ] This change is worth documenting at https://docs.all-hands.dev/
- [x] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality this introduces.**

- Make tracking for metrics more accurate and fault tolerant (accumulated cost, token usage, etc)
- Does not track metrics in a centralized manner yet (metrics reported in FE are for the main agent, and do not include metrics for auto generate convo titles, or running the condenser)

#### Problems

We have many problems with the way we handle budget caps in the agent controller

1. The budget information is reset when the user decides to continue after hitting the convo budget cap
2. We store accumulated cost in both the agent state and llm state, and they have eventual consistency
3. We extended budget when the traffic control state is `PAUSED`; but nowhere in the agent controller do we set it that way, and thus I'm unable to extend budget even if I wanted to
4. Throttle logic and iteration extension logic seem to be scattered across different parts of the agent controlled
5. When budget is extended, we simply extend the current max by the initial amount; we should use the current accumulated cost as the new starting point
6. When the agent is delegated, the initial amount is initialized as the max_budget setting. This means that the delegate extends the budget by twice the normal amount the parent would


#### Fixes

This PR we do the following

1. Removing any resetting of accumulated costs
2. Extend the budget when we are throttled, and remove the unnecessary complexity of `PAUSED` state for traffic control
3. Organize throttling and max cap extensions in one place (when the agent goes from error to running)
4. We extend the agent budget based on the current accumulated cost and the initial delta amount
5. We track delta amounts for budget now. This means if the agent or the parent hits the budget cap, we extend by a fixed delta amount (fixes delegate from extending budget cap in multiples, see [comment](https://github.com/All-Hands-AI/OpenHands/pull/8764#issuecomment-2919777095))
6. Controller state always holds the most current budget cap and accumulate cost, simplifying tracking

This helps us establish the following assumptions going forward

1. Accumulated cost is never reset, we're always looking at the current cost for the conversation even after extending budget
2. When restarting a conversation, we will resume with the same accumulated cost and budget caps from when we last left off
3. The metrics for the controller state and llm metrics are always the same. Metrics are always accumulated over time, we never start a fresh copy. Instead, for delegate agents we'll create a snapshot of the parent's metrics. The different between child and parent metrics gives us the local metrics for the delegate (this also fixes the problem where the controller metrics had eventual consistency with global llm metrics)


Fixes #8858
Fixes #8238

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:922b0a6-nikolaik   --name openhands-app-922b0a6   docker.all-hands.dev/all-hands-ai/openhands:922b0a6
```

pcuci commented on openhands/controller/state_manager.py:
```diff
@@ -0,0 +1,210 @@
+from openhands.events.stream import EventStream
+
+
+class StateManager:
```
this might make it easier to later extend to a use-case where users control the budget for each thread.

"to resolve this issue, don't spend more than $3; plan accordingly"

enyst commented on openhands/llm/metrics.py:
```diff
@@ -193,22 +193,6 @@ def get(self) -> dict:
             'token_usages': [usage.model_dump() for usage in self._token_usages],
         }
 
-    def reset(self) -> None:
```
I agree, let's kill this, and if we really need it, hey, we'll find out exactly why and can rethink.

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -600,40 +568,33 @@ async def set_agent_state_to(self, new_state: AgentState) -> None:
+            # Reset throttle state
+            self.state.traffic_control_state = TrafficControlState.NORMAL
+
+        if self._pending_action is not None and (
```
Is this `if self._pending_action is not None` new in this PR? If yes, why was it needed?

malhotra5 replied:
It's not new in this PR, not sure what it was being used for before

I moved this conditional from `elif` to `if` statement because the conditions for success are mutually exclusive

enyst commented on openhands/controller/state/deprecated_state.py:
```diff
@@ -0,0 +1,76 @@
+
+
+@dataclass
+class DeprecatedState:
```
🤔  That was intended, I'm pretty sure. Pickling stores the object, and restores it, the python object with possibly complex structure.

There may be some compatibility code needed when unpickling old objects, like for example in Metrics I think we make sure that the attributes exist (because some did not exist in the past).

malhotra5 replied:
I see warning like this when restoring the old agent_states files into the new schema

> 17:05:40 - openhands:WARNING: agent_session.py:490 - State could not be restored: Can't get attribute 'TrafficControlState' on <module 'openhands.controller.state.state' from '/Users/rohitmalhotra/Documents/Openhands/Openhands/openhands/controller/state/state.py'>

I think the pickle doesn't handle mismatches in attributes 

Will have to hold a deprecated copy for now; a way to mitigate this in the future is to serialize the class to a json first, then pickle the json

Ideally that could be a follow up to this PR

malhotra5 replied:
Yep we have default values for new fields

Also when restoring states with deprecated fields like `iteration`, we'll migrate these values to the new fields like `IterationControlFlag(current_value=deprecation_iteration_field_value)`

enyst commented on openhands/controller/agent_controller.py:
```diff
@@ -451,6 +451,8 @@ async def _handle_observation(self, observation: Observation) -> None:
             log_level, str(observation_to_print), extra={'msg_type': 'OBSERVATION'}
         )
 
+        # FIX ME: it seems like observations don't carry llm_metrics anymore; rather the LLM class aggregates cost
```
I think obs coming from the LLM draft editor were carrying metrics and should still do.

It's because that happens in the runtime... 😅 so we had no `state` there, and the `llm` is not the same, it's the "draft_editor" LLM.

We have too much complexity 😂

malhotra5 replied:
Ahh thanks for the clarification

I'll fix this in a follow up PR; this PR is centralizing all metrics accumulation in the controller.agent.llm object

In a future PR I'll

1. create exactly one llm object
2. pass this object around for all microservices (condensor, llm draft editor, etc)
3. this pr ensures that any updates to this centralized llm object will be reflected in the agent controller and budget control flags

malhotra5 replied:
^ i'll rewrite the comment here

enyst replied:
Not sure about the above. Having multiple LLMs working on different things is intentional, it makes a lot of sense to do editing application with special editing models, or simpler models, rather than the same LLM, for example.

I was going to suggest instead that, on the draft editor - to not merge its results, keep them separate and save them separately, idk, maybe with `state.draft_metrics` or something. Then the evals would update their scripts to compute total costs. 🤔

malhotra5 replied:
Yeah I think we'd need to discuss more, but I'd like to do that outside of this PR. This PR doesn't modify the way multiple services construct their own LLM objects atm.

My initial thoughts are that we should have total conversation cost to be centralized in one place. The way I see it is, having separate LLM objects would help with traceability for spend on individual services, but it requires aggregating them together later to have total spend. I'd rather it be one centralized metrics object, where we identify which service is attempting to accumulate more spend (auto generation of convo titles, draft editor, condenser, etc)

This way we accumulate total cost in one place but can still break it down in terms of spend per service if needed. We don't sacrifice on traceability but still have a singular source of truth for total convo cost. 

I'd prefer always having total convo and spend extra effort breaking it down by individual services, rather than having cost per service and having to put extra effort into accumulating them. We'd also prevent having fragmented stores for metrics, and don't have the burden of deciding where and how to store them when adding new services

malhotra5 replied:
^ oh i see the problem is having a different a model all together for the llm (like a small more lightweight one)

That should be possible still, I'm only talking about centralizing the metrics object  - I should clarify that in the comment

enyst replied:
It seems that github ate my comment for some reason. 🫣

I only wanted to note that this was a great argument:
> We'd also prevent having fragmented stores for metrics, and don't have the burden of deciding where and how to store them if we decide introduce more specialized services that require llm completions

and makes sense. (yes, the problem was the models; not the metrics, the metrics sorely need a rethink)

malhotra5 commented:
I think this is a good stopping point

1. we have fixed metrics management
2. since `AgentController` is a god object, we've introduced control flags and moved history management out of the controller

I'm afraid attempting to further simplify the agent controller from this point is out of scope for this PR

csmith49 commented on openhands/controller/state/control_flags.py:
```diff
@@ -0,0 +1,110 @@
+
+    def expand(self, headless_mode) -> None:
+        """Expand the budget limit by adding the initial value to the current value."""
+        if self._hit_limit:
```
Should this reference `self.reached_limit` instead of `self._hit_limit`? Any danger of accidentally using a stale `_hit_limit` value here?

Ditto for the `IterationControlFlag`.

malhotra5 replied:
Hmm good point

Prior to this PR we had throttleflag that would check if we are in a throttled state. This is so that we don't accidentally increase limits without user consent. 

Without throttle flags we can extend budget without consent if the user executes the last available step from their budget, hits an agent error state, and sends another message switching the agent back to a running state.

Whenever we switch from agent error to agent running, we attempt to expand the flag. The `_hit_limit` is to verify that we have thrown an error that the user sees before we attempt the expand the flag.

> Any danger of accidentally using a stale _hit_limit value here?

I should definitely write unit tests for this good point

csmith49 commented on openhands/controller/state/state.py:
```diff
@@ -75,35 +79,43 @@ class State:
-    max_iterations: int = 100
+    iteration_flag: IterationControlFlag = field(
+        default_factory=lambda: IterationControlFlag(
+            initial_value=100, current_value=0, max_value=100
```
`initial_value` is the amount the control flag gets expanded by, right? At least in this context it reads like we're setting the iteration flag already tripped.

enyst replied:
🤔 Could we name it... `step_increase`, if it's a step increase?

malhotra5 commented on openhands/controller/state/control_flags.py:
```diff
@@ -0,0 +1,110 @@
+    initial_value: T
+    current_value: T
+    max_value: T
+    headless_mode: bool = False
```
Yep its around so that we can migrate legacy agent states (which contain these values) to the new state schema (with one with IterationControlFlag and BudgetControlFlag)

malhotra5 replied:
Yep headless hits errors

UI you have the option to expand when user sends message

All of the logic is centralized and executed by the control flags

csmith49 commented on openhands/llm/metrics.py:
```diff
@@ -221,5 +205,58 @@ def copy(self) -> 'Metrics':
+        base_usage = baseline.accumulated_token_usage
+        current_usage = self.accumulated_token_usage
+
+        result._accumulated_token_usage = TokenUsage(
+            model=self.model_name,
+            prompt_tokens=current_usage.prompt_tokens - base_usage.prompt_tokens,
+            completion_tokens=current_usage.completion_tokens
+            - base_usage.completion_tokens,
+            cache_read_tokens=current_usage.cache_read_tokens
+            - base_usage.cache_read_tokens,
+            cache_write_tokens=current_usage.cache_write_tokens
+            - base_usage.cache_write_tokens,
+            context_window=current_usage.context_window,
+            per_turn_token=0,
+            response_id='',
+        )
```
Suggestion for future refactors: we can add `__sub__` to `TokenUsage` which does all this arithmetic for us with `current_usage - base_usage`.

malhotra5 replied:
Great point, will do in a follow up pr

csmith49 commented on openhands/controller/state/state_tracker.py:
```diff
@@ -0,0 +1,248 @@
+
+
+class StateTracker:
+    """ """
```
Can you add some of the context from the discussion with Engel to this doc-string?

malhotra5 commented:
@enyst would like that I hold off from merging util you've had more time to look into this?

enyst commented:
I think we’re better off merging, we need this. Thank you!

malhotra5 commented:
Thanks for the help with reviews! I did one last QA and it went well, will be merging
